### PR TITLE
Add barrier-free collective algorithms

### PR DESCRIPTION
**Pull request by chhwang**

### CODE REVIEW

File: test/mscclpp-test/allgather_test.cu [hunk 1]
```diff
@@ -288,6 +288,49 @@ __global__ void allgather4(int rank, int worldSize, int nRanksPerNode, size_t ne
...
+  const int wid = tid / WARP_SIZE;
+  const int nWarp = blockDim.x * gridDim.x / WARP_SIZE;
+  const int nPeer = nRanksPerNode - 1;
+  const int chanOffset = nPeer * blockIdx.x;
+  auto smChans = constSmChans + chanOffset;
+
+  if (wid < nPeer) {
+    smChans[wid].signal();
+    smChans[wid].wait();
+  }
```
Comment (Binyang2014) [inline test/mscclpp-test/allgather_test.cu:303]: Do we need all threads in a warp to call signal/wait? Or just need one thread in a warp
```cpp
if (wid < nPeer and lid == 0)
```

File: test/mscclpp-test/allgather_test.cu [hunk 2]
```diff
@@ -288,6 +288,49 @@ __global__ void allgather4(int rank, int worldSize, int nRanksPerNode, size_t ne
...
+  const size_t unitBytes = unitBytesPerWarp * nWarp;
+  const size_t bytesPerGPU = nelemsPerGPU * sizeof(int);
+  const size_t bytes = bytesPerGPU * nPeer;
+  const size_t nLoop = bytes / unitBytes;
+  for (size_t i = 0; i < nLoop; ++i) {
+    const size_t gWid = wid + i * nWarp;
+    const int peerIdx = gWid % nPeer;
+    const int remoteRankLocalIndex = (peerIdx < rank ? peerIdx : peerIdx + 1);
+    const size_t offset = bytesPerGPU * remoteRankLocalIndex + (gWid / nPeer) * unitBytesPerWarp;
+    smChans[peerIdx].get(offset, unitBytesPerWarp, lid, WARP_SIZE);
```
Comment (Binyang2014) [inline test/mscclpp-test/allgather_test.cu:316]: I am a bit confused about this algo. For example, a thread block has 16 warps. Then the peerIdx is: "0,1,2,3,4,5,6,0,1,2,3,4,5,6,0,1". As you see they will get 3 pieces of data from peer `0,1`, and two pieces of data from `"0,1,2,3,4,5,6`. Seems not correct.

File: test/mscclpp-test/allgather_test.cu (hunk 1 above)
Comment (chhwang) [reply test/mscclpp-test/allgather_test.cu]: @Binyang2014 Thank you for finding this! Fixed in [cc9c150](https://github.com/microsoft/mscclpp/pull/248/commits/cc9c1505207848efd1d7d68dcaddac48387142a7).

**Comment by chhwang** [comment]
Moved to https://github.com/microsoft/mscclpp/pull/256

**Pull request closed without merging**